src/shared/redirect.ts

```diff
@@ -0,0 +1,35 @@
+import { NextPageContext } from "next";
+import absolute from "next-absolute-url";
+import { getUserId } from "./session";
+
+export const redirect = async (ctx: NextPageContext) => {
+  const userId = await getUserId(ctx);
+  const { req } = ctx;
+
+  if (!req) {
+    return {
+      redirect: {
+        destination: "/404",
+        permanent: false,
+      },
+    };
+  }
+
+  const { url } = req;
```

Comment:
Use resolvedUrl to avoid strange data fetching urls coming through with SSR.

According to the [Next docs](https://nextjs.org/docs/basic-features/data-fetching#getserversideprops-server-side-rendering), resolvedUrl is
> A normalized version of the request URL that strips the _next/data prefix for client transitions and includes original query values.
> 
```suggestion
  const url = ctx.resolvedUrl;
```